### ISSUE TITLE
manager: improve error reporting in m4dapplication status

### ISSUE THREAD

**Issue by shlomitk1** [IBM/the-mesh-for-data#67]
The goal is to revisit the error management done by M4DApplication controller.
Proposal: 
M4DApplication status includes two conditions: Error and Failure.
When a non-fatal error is received (e.g. grpc communication error), an Error condition becomes active and another retrial is attempted by returning an error in reconcile.
When a fatal error is received, a Failure condition becomes active and no retrials are made.

**Comment by roee88** [comment]
Apologies for commenting so late and only after a PR is opened. We talked about error handling so many times but never documented something that the entire project should follow. I wonder if we need to do that first or should we move forward as is and open a new issue, understanding that the changes proposed here might even be reverted.

This issue focuses on M4DApplication status so I will focus on it and ignore the current implementation in Blueprint. 

> Error and Failure conditions are misleading. Staying with a single error string may simplify things.

There are two statements here:  

One is to stop using error conditions. Are you claiming that not using conditions simplifies the codebase, the usage, or both?

The second is to remove the field-level separation between possibly transient operational errors and fatal errors, but just be aware that we should reflect this information in some way (even if within a message).

> Some errors are critical

Can we map when things are critical and when not? My understanding is that critical is anything that requires one of the following before we can try again:
1. M4DApplication spec to change
1. Policies to change
1. M4D operator/admin need to add resources (storage, compute, modules)
1. M4D operator/admin need to change configuration (e.g., vault token)

Examples for non critical are:
1. Communication errors with vault
1. Communication errors with connectors
1. Errors from connectors
1. Kubernetes errors on apply/update/get (maybe w/ some exceptions)

> , others should result in a number of retrials up to some limit, e.g. communication problems.

Why should we create a max limit or handle the retries on our own? controller-runtime already [uses](https://github.com/kubernetes/client-go/blob/77eda6a9395bd7884e99d484f29ae3c6d55f0eb9/util/workqueue/default_rate_limiters.go#L39) an exponential backoff algorithm on errors. The model should be that if there is a transient error then at some point it's fixed without a human touching it.

> Some errors are not informative enough.

M4DApplication is mostly intended for users and not operators (cf. Blueprint). There are two approaches here:
1. Contain all error details even if the one creating the M4DApplication would be clueless about what it means because it reflects components that he shouldn't even be aware of.
2. Contain error details that are meaningful for users, whereas Blueprint errors are useful for operators to further debug operational issues 

I personally would go for the latter. In that case, we should understand what is useful for users and map to it, possibly in the future with suggestions on what the user can do to fix it (even if it's "go to operator" in some cases). Some made up examples to demonstrate what I mean:
1. Access if forbidden because you can't access X (go to governance officer if you think...)
1. System lacks storage resources (go to operator)
1. System lacks modules (change X in spec or go to operator if you think...)
1. Operational error (summary of error, go to operator if it persists)

**Comment by shlomitk1** [comment]
@roee88 
1. The problem with the current conditions is that they do not reflect the real state to the user. Either we use staged conditions, such as BlueprintGenerated (yes/no, errors), Ready (yes/no, errors, usage instructions), or stay with a boolean Ready and an error to report in case it will never become ready. 
2. The separation between Error and Failure was problematic when what we really want is to know when to stop and when to retry. Also, if we retry then the user does not have to know about the error.
The PR attempts to eliminate the existing conditions and retry in some cases. The errors leading to retrials are various communication errors (not including invalid argument). Resource update errors trigger reconcile as well. 
Informative error messages - work is still in progress. For now, it is easier to collect as many problems as possible, for the debug purposes. As for the users, the errors may be either "access is forbidden", in which case getting more explanation from the connectors would be helpful, or "configuration problems - go to operator". The PR only did some unified restructuring of the message, adding the relevant asset ID, the service/component that originated the error.
Number of retrials - are you saying that a simple Requeue: true will cause reconcile again and again with a longer time interval?  It would indeed be better this way.

**Comment by roee88** [comment]
> The problem with the current conditions is that they do not reflect the real state to the user. Either we use staged conditions, such as BlueprintGenerated (yes/no, errors), Ready (yes/no, errors, usage instructions), or stay with a boolean Ready and an error to report in case it will never become ready.

What's wrong with having an error condition? How is it different? (except for being shown nicely in dashboards)

> Also, if we retry then the user does not have to know about the error.

I disagree. As a user I would like to know that something is wrong even if it might get automatically fixed at some point. I don't need full details about components that I shouldn't even be aware about, but I do want to know the type of issue.

> Number of retrials - are you saying that a simple Requeue: true will cause reconcile again and again with a longer time interval? It would indeed be better this way.

If error is returned, regardless of result

**Comment by shlomitk1** [comment]
It does not really matter - condition or a string, having both Failure and Error conditions seemed artificial. I can change it back to a condition. However, having removed maximum retrials, I am now missing a way to turn a temporary problem into a failure that must be reported. A solution might be to do as you suggested: report problems in any case, while trying to re-attempt in some cases. But then we need to inform the user in some way, for example: "An error has occurred during communication with xxx. If the error persists, please contact the operator. Description: <error>"

**Comment by roee88** [comment]
You can add a Fatal boolean to the condition maybe?

**Comment by shlomitk1** [comment]
> 
> 
> You can add a Fatal boolean to the condition maybe?

Which brings us back to Failure vs. Error conditions...

**Comment by roee88** [comment]
I really don't get the problem and I think that you misunderstood me

1. We want users to be aware that there are issues even if possibly transient (immediately, don't hide it from me)
2. We want to retry on transient errors until they get resolved but not on fatal ones (it can take some time, we never quit, exponential backoff is good enough)
3. We want the user to know if this is a transient error or not (you suggested via the error text but it can also be a flag, boolean only condition or whatever)

What is complex or artificial here?

**Comment by shlomitk1** [comment]
> 
> 
> I really don't get the problem and I think that you misunderstood me
> 
>     1. We want users to be aware that there are issues even if possibly transient (immediately, don't hide it from me)
> 
>     2. We want to retry on transient errors until they get resolved but not on fatal ones (it can take some time, we never quit, exponential backoff is good enough)
> 
>     3. We want the user to know if this is a transient error or not (you suggested via the error text but it can also be a flag, boolean only condition or whatever)
> 
> 
> What is complex or artificial here?

It's not, just a reflection on how it came back to the original concept.
It was artificial in the old codebase because both conditions were handled basically in the same way.
I'll rearrange the code accordingly.
